File: algorithm + contest/pattern/Trie.py
from typing import List


class TrieNode:
    def __init__(self):
        self.children = [None] * 26  # [a-z]
        self.is_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, val: str):

        node = self.root
        for c in val:
            pos = ord(c) - ord('a')

            if self.children[pos] is None:
                self.children[pos] = TrieNode()

            node = self.children[pos]

        node.is_word = True

    def search(self, word: str) -> bool:
        node = self.root

        for ch in word:
            idx = ord(ch) - ord('a')

            if node.children[idx] is None:
                return False

            node = node.children[idx]

        return node.is_word

    def startsWith(self, prefix: str) -> bool:
        node = self.root

        for ch in prefix:
            idx = ord(ch) - ord('a')

            if node.children[idx] is None:
                return False

            node = node.children[idx]

        return True


class ReverseTrie:
    def __init__(self, words: List[str]):
        self.root = TrieNode()
        self.stream_characters = []

        for word in words:
            node = self.root
            for c in reversed(word):
                pos = ord(c) - ord('a')
                if node.children[pos] is None:
                    node.children[pos] = TrieNode()
                node = node.children[pos]
            node.is_word = True

    def query(self, letter: str) -> bool:
        self.stream_characters.append(letter)

        node = self.root
        for c in reversed(self.stream_characters):
            pos = ord(c) - ord('a')
            if node.is_word:
                return True
            if node.children[pos] is None:
                return False
            node = node.children[pos]

        return node.is_word

# example
main = ReverseTrie(["cd", "f", "kl"])
print(main.query("a"))
print(main.query("b"))
print(main.query("c"))
print(main.query("d"))
print(main.query("e"))
print(main.query("f"))
print(main.query("g"))
print(main.query("h"))
print(main.query("i"))
print(main.query("j"))
print(main.query("k"))
print(main.query("l"))
print(main.query("m"))
print(main.query("n"))
print(main.query("o"))
print(main.query("p"))
print(main.query("q"))
print(main.query("r"))
print(main.query("s"))
print(main.query("t"))
print(main.query("u"))
print(main.query("v"))
print(main.query("w"))
print(main.query("x"))
print(main.query("y"))
print(main.query("z"))

File: interview/leetcode/day07012026/Trie.java
// The most imprortant part to grasp is that node does not store whole string.
// Instead, each node represents a single character in the string, and the path from the root to 
// a specific node represents a prefix of the string or the whole string itself (with flag)
public class Trie {

  // If we don't use static here, TrieNode will be inner class of Trie,
  // and each TrieNode instance will hold a reference to the outer Trie instance,
  // which is unnecessary and consumes extra memory.
  private final TrieNode root;

  static class TrieNode {
    TrieNode[] tree;
    boolean isWord;

    TrieNode() {
      tree = new TrieNode[26]; // assuming only lowercase a-z
      isWord = false;
    }
  }

  public Trie() {
    root = new TrieNode();
  }

  public void insert(String word) {
    TrieNode curr = root;
    for (char c : word.toCharArray()) {
      if (curr.tree[c - 'a'] == null) {
        curr.tree[c - 'a'] = new TrieNode();
      }
      curr = curr.tree[c - 'a'];
    }

    curr.isWord = true;
  }

  public boolean search(String word) {
    TrieNode node = findWord(word);
    return node != null && node.isWord; // it can be not a word, just a prefix. Mission is search word
  }

  public boolean startsWith(String word) {
    return findWord(word) != null;
  }

  public TrieNode findWord(String word) {
    TrieNode curr = root;
    for (char c : word.toCharArray()) {
      if (curr.tree[c - 'a'] == null) {
        return null;
      }

      curr = curr.tree[c - 'a'];
    }

    return curr;
  }

  public static void main(String[] args) {
    Trie trie = new Trie();
    trie.insert("apple");
    System.out.println(trie.search("apple")); // true
    System.out.println(trie.search("app")); // false
    System.out.println(trie.startsWith("app")); // true
    trie.insert("app");
    System.out.println(trie.search("app")); // true
  }
}
